src/DataProtection/custom/Cmdlets/Platform/Vault/Set-AzDataProtectionMSIPermission.ps1

```diff
@@ -0,0 +1,205 @@
+                      }
+
+                      elseif($PermissionsScope -eq "Subscription")
+                      { 
+                          AssignMissingRoles -ObjectId $vault.IdentityPrincipalId -Permission $Permission -ResourceScope $KeyvaultSubscriptionName
+                      }                 
+
+                      Write-Host "Assigned Required RBAC Permissions"
+                  }
+              }
+          }
+               
+          foreach($Permission in $manifest.datasourcePermissions)
+          {
+              $CheckPermission = $AllRoles
+              | Where-Object { ($_.Scope -eq $DataSourceId -or $_.Scope -eq $ResourceRG -or  $_.Scope -eq $SubscriptionName) -and $_.RoleDefinitionName -eq $Permission}
+              
+              if($CheckPermission -ne $null)
+              {
+                  Write-Host "Required Resource Permissions Already Assigned."
+              }
+
+              else
+              {
+                  $MissingRolesInitially = $true
+
+                  Write-Host "Assigning Required Resource Permissions"
+
+                  if($PermissionsScope -eq "Resource")
+                  {
+                     AssignMissingRoles -ObjectId $vault.IdentityPrincipalId -Permission $Permission -ResourceScope $DatasourceId
+                  }
+              
+                  elseif($PermissionsScope -eq "ResourceGroup")
+                  {
+                     AssignMissingRoles -ObjectId $vault.IdentityPrincipalId -Permission $Permission -ResourceScope $ResourceRG
+                  }
+
+                  elseif($PermissionsScope -eq "Subscription")
+                  {                   
+                     AssignMissingRoles -ObjectId $vault.IdentityPrincipalId -Permission $Permission -ResourceScope $SubscriptionName
+                  }
+
+                  Write-Host "Assigned Required Resource Permissions"
+              }
+          }
+              
+
+          foreach($Permission in $manifest.snapshotRGPermissions)
+          {
+              $SnapshotResourceGroupId = $BackupInstance.Property.PolicyInfo.PolicyParameter.DataStoreParametersList[0].ResourceGroupId
+
+              $CheckPermission = $AllRoles
+              | Where-Object { ($_.Scope -eq $SnapshotResourceGroupId -or $_.Scope -eq $SubscriptionName)  -and $_.RoleDefinitionName -eq $Permission}
+
+              if($CheckPermission -ne $null)
+              {
+                  Write-Host "Required SnapshotRG Permissions Already Assigned."
+              }
+
+              else
+              {
+                  $MissingRolesInitially = $true
+
+                  Write-Host "Assigning Required SnapshotRG Permissions"
+
+                  if($PermissionsScope -eq "Subscription")
+                  {
+                      AssignMissingRoles -ObjectId $vault.IdentityPrincipalId -Permission $Permission -ResourceScope $SubscriptionName
+                  }             
+                  else
+                  {
```

Comment:
This code can be shortened with use of a helper function that on providing the armid and the scope returns the corresponding scope id.